source/base/Text.ooc

```diff
@@ -0,0 +1,298 @@
+				start += 2
+				lastValidIndex += 2
+			}
+			for (i in start .. this count) {
+				if (This _isNumeric(this[i], base))
+					++lastValidIndex
+				else
+					break
+			}
+			power := 1 as ULong
+			while (lastValidIndex >= 0) {
+				result += power * This _toInt(this[lastValidIndex], base)
+				--lastValidIndex
+				power *= base
+			}
+		}
+		result
+	}
+	toFloat: func -> Float {
+		this toLDouble() as Float
+	}
+	toDouble: func -> Double {
+		this toLDouble() as Double
+	}
+	toLDouble: func -> LDouble {
+		result := 0.0 as LDouble
+		if (!this isEmpty) {
+			sign := 1
+			start := 0
+			if (this[0] == '-') {
+				start = 1
+				sign = -1
+			}
+			index := this count
+			for (i in start .. this count) {
+				if (!This _isNumeric(this[i], 10)) {
+					index = i
+					break
+				}
+			}
+			result = this slice(start, index - start) toLLong~inBase(10)
+			if (index > -1 && index < this count) {
+				if (this[index] == '.') {
+					power := 0.1
+					for (i in index + 1 .. this count) {
+						if (This _isNumeric(this[i], 10)) {
+							result += power * This _toInt(this[i], 10)
+							power /= 10
+						} else {
+							index = i
+							break
+						}
+					}
+				}
+				if (this[index] == 'e' || this[index] == 'E') {
+					exponent := this slice(index + 1, this count - index) toInt~inBase(10)
+					result = result * pow(10, exponent)
+				}
+			}
+			result *= sign
+		}
+		result
+	}
+	_allocateAndCopy: func@ (source: CString, size: Int) {
+		this _count = size
+		this _buffer = gc_malloc(size)
+		memcpy(this _buffer, source, size)
+	}
+	_changeOwnership: func@ (target: Owner) -> Bool {
+		result: Bool
+		if (result = (target != this _owner && target != Owner Stack && target != Owner Literal)) {
+			if (result = (this _owner == Owner Stack || this _owner == Owner Literal)) {
+				source := this _buffer
+				this _allocateAndCopy(source, this count)
+			}
+			this _owner = target
+		}
+		result
+	}
+	_detectNumericBase: func -> Int {
+		result := 10
+		if (this beginsWith(This new(c"0x", 2)))
+			result = 16
+		result
+	}
+	_isNumeric: static func (c: Char, base: Int) -> Bool {
+		result := (c >= '0') && (c <= '9')
+		if (!result && base == 16)
+			result = (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F')
+		result
+	}
+	_toInt: static func (c: Char, base: Int) -> Int {
+		result := 0
+		if (_isNumeric(c, base)) {
+			if (base != 16)
+				result = (c - '0') as Int
+			else {
+				if (c >= 'a' && c <= 'f')
+					result = (c - 'a') as Int + 10
+				else if (c >= 'A' && c <= 'F')
+					result = (c - 'A') as Int + 10
+				else if (c >= '0' && c <= '9')
+					result = (c - '0') as Int
+			}
+		}
+		result
+	}
+	empty: static This { get { This new() } }
+}
+
+operator == (left: String, right: Text) -> Bool { Text new(left) == right }
+operator != (left: String, right: Text) -> Bool { !(left == right) }
+operator != (left: Text, right: String) -> Bool { !(left == right) }
+operator != (left: Text, right: Text) -> Bool { !(left == right) }
```

Comment:
This is a very long file... and some function are located far away from where they are called, only because they're private and the calling functions are not. I don't see an easy way to fix that, though.